frontends/openssl-provider/common.c

```diff
@@ -0,0 +1,406 @@
+}
+
+static int esdm_rand_uninstantiate(void *ctx __unused)
+{
+  return 1;
+}
+
+static int esdm_rand_generate(void *ctx __unused, unsigned char *out,
+                              size_t outlen, unsigned int strength __unused,
+                              int prediction_resistance,
+                              const unsigned char *addin, size_t addin_len)
+{
+  ssize_t ret;
+
+  if (!out)
+    goto err;
+
+  if (addin && addin_len > 0) {
+    esdm_invoke(esdm_rpcc_write_data(addin, addin_len));
+    if (ret != 0)
+      goto err;
+  }
+
+  if (prediction_resistance) {
+    esdm_invoke(esdm_rpcc_get_random_bytes_pr(out, outlen));
+  } else {
+    esdm_invoke(esdm_rpcc_get_random_bytes_full(out, outlen));
+  }
+  if (ret != (ssize_t)outlen)
+    goto err;
+
+  return 1;
+
+err:
+  return 0;
+}
+
+static int esdm_rand_reseed(void *ctx __unused,
+                            int prediction_resistance __unused,
+                            const unsigned char *ent, size_t ent_len,
+                            const unsigned char *addin, size_t addin_len)
+{
+  ssize_t ret;
+
+  /* unaccounted writing of additional data does no harm */
+  if (ent && ent_len > 0)
+    esdm_invoke(esdm_rpcc_write_data(ent, ent_len));
+  if (addin && addin_len > 0)
+    esdm_invoke(esdm_rpcc_write_data(addin, addin_len));
+
+  return 1;
+}
+
+static size_t esdm_rand_nonce(void *ctx __unused, unsigned char *out,
+                              unsigned int outlen, size_t min_noncelen __unused,
+                              size_t max_noncelen __unused)
+{
+  ssize_t ret;
+
+  if (out == NULL)
+    return outlen;
+
+  esdm_invoke(esdm_rpcc_get_random_bytes_full(out, outlen));
+
+  if (ret == (ssize_t)outlen)
+    return outlen;
+  else
+    return 0;
+}
+
+static size_t esdm_rand_get_seed(void *ctx __unused, unsigned char **buffer,
+                                 int entropy, size_t min_len __unused,
+                                 size_t max_len __unused,
+                                 int prediction_resistance __unused,
+                                 const unsigned char *addin, size_t addin_len)
+{
+  struct esdm_seed_buffer {
+    uint64_t len;
+    uint64_t entropyBits;
+    uint8_t buf[2048] /* should be large enough for entropy from all active
+                         sources */
+  } __attribute__((__packed__));
+  size_t buffer_size = sizeof(struct esdm_seed_buffer);
+  struct esdm_seed_buffer *seed_buf;
+  ssize_t ret;
+
+  *buffer = OPENSSL_secure_zalloc(sizeof(struct esdm_seed_buffer));
+
+  if (sizeof(struct esdm_seed_buffer) < min_len)
+    goto err;
+  if (sizeof(struct esdm_seed_buffer) >= max_len)
+    goto err;
+
+  if (addin && addin_len > 0) {
+    esdm_invoke(esdm_rpcc_write_data(addin, addin_len));
+    if (ret != 0)
+      goto err;
+  }
+
+  esdm_invoke(esdm_rpcc_get_seed(*buffer, buffer_size, 0));
+  if (ret <= 0)
+    goto err;
+
+  seed_buf = (struct esdm_seed_buffer *)*buffer;
+  if (seed_buf->entropyBits < (uint64_t)entropy)
```

Comment:
Sanity check: are you sure that the OpenSSL entropy is measured in bits here? If yes, may I ask you to rename the variable to entropy_bits.

Side note: I lost days in debugging issues of bits vs bytes. Thus, I made a habit now to mark all variables that hold bit sizes.

Reply:
I'm pretty sure, that the entropy argument uses bits. I'll rename the argument accordingly. It is stated in the documentation and I also found it in the OpenSSL source code, e.g.:

```c
static size_t seed_get_seed(void *vseed, unsigned char **pout,
                            int entropy, size_t min_len, size_t max_len,
                            int prediction_resistance,
                            const unsigned char *adin, size_t adin_len)
{
    size_t bytes_needed;
    unsigned char *p;

    /*
     * Figure out how many bytes we need.
     * This assumes that the seed sources provide eight bits of entropy
     * per byte.  For lower quality sources, the formula will need to be
     * different.
     */
    bytes_needed = entropy >= 0 ? (entropy + 7) / 8 : 0;
    if (bytes_needed < min_len)
        bytes_needed = min_len;
    if (bytes_needed > max_len) {
        ERR_raise(ERR_LIB_PROV, PROV_R_ENTROPY_SOURCE_STRENGTH_TOO_WEAK);
        return 0;
    }

    p = OPENSSL_secure_malloc(bytes_needed);
    if (p == NULL)
        return 0;
    if (seed_src_generate(vseed, p, bytes_needed, 0, prediction_resistance,
                          adin, adin_len) != 0) {
        *pout = p;
        return bytes_needed;
    }
    OPENSSL_secure_clear_free(p, bytes_needed);
    return 0;
}
```